{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a3678307",
   "metadata": {},
   "source": [
    " #               Exception Handling Assignment II\n",
    " Bhumanyu_Singh_PW_DS_2_Assignment_13_feb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ae42d19",
   "metadata": {},
   "source": [
    "#### Q1. Explain why we have to use the Exception class while creating a Custom Exception.\n",
    "Note: Here Exception class refers to the base class for all the exceptions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b9c948c",
   "metadata": {},
   "source": [
    "In python there is inbuilt Exception is created which will help us to create costomised execption class as a child class of an inbuilt exveption"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d8b8405",
   "metadata": {},
   "source": [
    "In Python, when we create a custom exception, we typically inherit from the built-in Exception class. The Exception class serves as the base class for all the exceptions in Python, and it provides a set of methods and attributes that are essential for proper exception handling.\n",
    "\n",
    "Here are the reasons why we use the Exception class as the base class for custom exceptions:\n",
    "\n",
    "Consistency: By inheriting from the Exception class, our custom exception becomes consistent with other built-in exceptions in Python. This allows us to use our custom exception in a similar way as we would use any other exception in Python.\n",
    "\n",
    "Proper Exception Handling: The Exception class provides methods like __str__() and __repr__() that help in formatting the exception message and displaying useful information when the exception is raised. By inheriting from the Exception class, our custom exception can leverage these methods, making it easier for developers to handle and understand the exception.\n",
    "\n",
    "Customization: By creating a custom exception class, we can define additional attributes and methods that are specific to our application or use case. We can add extra context or information to the exception, making it more informative and helpful for debugging.\n",
    "\n",
    "Catching Specific Exceptions: When we raise a custom exception, we can catch it specifically using except blocks, which allows us to handle different exceptions differently based on their types. This level of granularity in exception handling is essential for writing robust and error-tolerant code.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ab9f4cd",
   "metadata": {},
   "source": [
    "#### Q2. Write a python program to print Python Exception Hierarchy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fd3619b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python Exception Hierarchy:\n",
      "BaseException\n",
      "    Exception\n",
      "        TypeError\n",
      "            FloatOperation\n",
      "            MultipartConversionError\n",
      "        StopAsyncIteration\n",
      "        StopIteration\n",
      "        ImportError\n",
      "            ModuleNotFoundError\n",
      "                PackageNotFoundError\n",
      "                PackageNotFoundError\n",
      "            ZipImportError\n",
      "        OSError\n",
      "            ConnectionError\n",
      "                BrokenPipeError\n",
      "                ConnectionAbortedError\n",
      "                ConnectionRefusedError\n",
      "                ConnectionResetError\n",
      "                    RemoteDisconnected\n",
      "            BlockingIOError\n",
      "            ChildProcessError\n",
      "            FileExistsError\n",
      "            FileNotFoundError\n",
      "            IsADirectoryError\n",
      "            NotADirectoryError\n",
      "            InterruptedError\n",
      "                InterruptedSystemCall\n",
      "            PermissionError\n",
      "            ProcessLookupError\n",
      "            TimeoutError\n",
      "            UnsupportedOperation\n",
      "            herror\n",
      "            gaierror\n",
      "            timeout\n",
      "            Error\n",
      "                SameFileError\n",
      "            SpecialFileError\n",
      "            ExecError\n",
      "            ReadError\n",
      "            SSLError\n",
      "                SSLCertVerificationError\n",
      "                SSLZeroReturnError\n",
      "                SSLWantReadError\n",
      "                SSLWantWriteError\n",
      "                SSLSyscallError\n",
      "                SSLEOFError\n",
      "            URLError\n",
      "                HTTPError\n",
      "                ContentTooShortError\n",
      "            BadGzipFile\n",
      "        EOFError\n",
      "            IncompleteReadError\n",
      "        RuntimeError\n",
      "            RecursionError\n",
      "            NotImplementedError\n",
      "                ZMQVersionError\n",
      "                StdinNotImplementedError\n",
      "            _DeadlockError\n",
      "            BrokenBarrierError\n",
      "            BrokenExecutor\n",
      "                BrokenThreadPool\n",
      "            SendfileNotAvailableError\n",
      "            ExtractionError\n",
      "            RemovedCommandError\n",
      "            PackageDiscoveryError\n",
      "            VariableError\n",
      "        NameError\n",
      "            UnboundLocalError\n",
      "        AttributeError\n",
      "            FrozenInstanceError\n",
      "        SyntaxError\n",
      "            IndentationError\n",
      "                TabError\n",
      "        LookupError\n",
      "            IndexError\n",
      "            KeyError\n",
      "                NoSuchKernel\n",
      "                UnknownBackend\n",
      "            CodecRegistryError\n",
      "        ValueError\n",
      "            UnicodeError\n",
      "                UnicodeEncodeError\n",
      "                UnicodeDecodeError\n",
      "                UnicodeTranslateError\n",
      "            UnsupportedOperation\n",
      "            JSONDecodeError\n",
      "            SSLCertVerificationError\n",
      "            Error\n",
      "            ClassNotFound\n",
      "            ClipboardEmpty\n",
      "            MessageDefect\n",
      "                NoBoundaryInMultipartDefect\n",
      "                StartBoundaryNotFoundDefect\n",
      "                CloseBoundaryNotFoundDefect\n",
      "                FirstHeaderLineIsContinuationDefect\n",
      "                MisplacedEnvelopeHeaderDefect\n",
      "                MissingHeaderBodySeparatorDefect\n",
      "                MultipartInvariantViolationDefect\n",
      "                InvalidMultipartContentTransferEncodingDefect\n",
      "                UndecodableBytesDefect\n",
      "                InvalidBase64PaddingDefect\n",
      "                InvalidBase64CharactersDefect\n",
      "                InvalidBase64LengthDefect\n",
      "                HeaderDefect\n",
      "                    InvalidHeaderDefect\n",
      "                    HeaderMissingRequiredValue\n",
      "                    NonPrintableDefect\n",
      "                    ObsoleteHeaderDefect\n",
      "                    NonASCIILocalPartDefect\n",
      "            IllegalMonthError\n",
      "            IllegalWeekdayError\n",
      "            MacroToEdit\n",
      "            ParserError\n",
      "            InvalidFileException\n",
      "            UnequalIterablesError\n",
      "            InvalidVersion\n",
      "            _InvalidELFFileHeader\n",
      "            InvalidWheelFilename\n",
      "            InvalidSdistFilename\n",
      "            InvalidSpecifier\n",
      "            InvalidMarker\n",
      "            UndefinedComparison\n",
      "            UndefinedEnvironmentName\n",
      "            InvalidRequirement\n",
      "                RequirementParseError\n",
      "            UnequalIterablesError\n",
      "            _InvalidELFFileHeader\n",
      "            InvalidVersion\n",
      "            InvalidWheelFilename\n",
      "            InvalidSdistFilename\n",
      "            InvalidSpecifier\n",
      "            InvalidMarker\n",
      "            UndefinedComparison\n",
      "            UndefinedEnvironmentName\n",
      "            InvalidRequirement\n",
      "        AssertionError\n",
      "        ArithmeticError\n",
      "            FloatingPointError\n",
      "            OverflowError\n",
      "            ZeroDivisionError\n",
      "                DivisionByZero\n",
      "                DivisionUndefined\n",
      "            DecimalException\n",
      "                Clamped\n",
      "                Rounded\n",
      "                    Underflow\n",
      "                    Overflow\n",
      "                Inexact\n",
      "                    Underflow\n",
      "                    Overflow\n",
      "                Subnormal\n",
      "                    Underflow\n",
      "                DivisionByZero\n",
      "                FloatOperation\n",
      "                InvalidOperation\n",
      "                    ConversionSyntax\n",
      "                    DivisionImpossible\n",
      "                    DivisionUndefined\n",
      "                    InvalidContext\n",
      "        SystemError\n",
      "            CodecRegistryError\n",
      "        ReferenceError\n",
      "        MemoryError\n",
      "        BufferError\n",
      "        Warning\n",
      "            UserWarning\n",
      "                GetPassWarning\n",
      "                FormatterWarning\n",
      "                _WouldIgnoreField\n",
      "                _ExperimentalProjectMetadata\n",
      "                _InvalidFile\n",
      "            DeprecationWarning\n",
      "                ProvisionalWarning\n",
      "            PendingDeprecationWarning\n",
      "            SyntaxWarning\n",
      "            RuntimeWarning\n",
      "                ProactorSelectorThreadWarning\n",
      "                UnknownTimezoneWarning\n",
      "                PEP440Warning\n",
      "            FutureWarning\n",
      "                ProvisionalCompleterWarning\n",
      "            ImportWarning\n",
      "            UnicodeWarning\n",
      "            BytesWarning\n",
      "            ResourceWarning\n",
      "            DeprecatedTzFormatWarning\n",
      "            SetuptoolsDeprecationWarning\n",
      "                DistDeprecationWarning\n",
      "            PkgResourcesDeprecationWarning\n",
      "        Error\n",
      "        _OptionError\n",
      "        error\n",
      "        Verbose\n",
      "        _Error\n",
      "        SubprocessError\n",
      "            CalledProcessError\n",
      "            TimeoutExpired\n",
      "        _GiveupOnSendfile\n",
      "        error\n",
      "        LZMAError\n",
      "        RegistryError\n",
      "        _GiveupOnFastCopy\n",
      "        ZMQBaseError\n",
      "            ZMQError\n",
      "                ContextTerminated\n",
      "                Again\n",
      "                InterruptedSystemCall\n",
      "            ZMQBindError\n",
      "            NotDone\n",
      "        Error\n",
      "        error\n",
      "        PickleError\n",
      "            PicklingError\n",
      "            UnpicklingError\n",
      "        _Stop\n",
      "        TokenError\n",
      "        StopTokenizing\n",
      "        ClassFoundException\n",
      "        EndOfBlock\n",
      "        TraitError\n",
      "        ArgumentError\n",
      "        ArgumentTypeError\n",
      "        ConfigError\n",
      "            ConfigLoaderError\n",
      "                ArgumentError\n",
      "            ConfigFileNotFound\n",
      "        ConfigurableError\n",
      "            MultipleInstanceError\n",
      "        ApplicationError\n",
      "        NoIPAddresses\n",
      "        Error\n",
      "            CancelledError\n",
      "            TimeoutError\n",
      "            InvalidStateError\n",
      "        Incomplete\n",
      "        TimeoutError\n",
      "        InvalidStateError\n",
      "        LimitOverrunError\n",
      "        QueueEmpty\n",
      "        QueueFull\n",
      "        InvalidPortNumber\n",
      "        Empty\n",
      "        Full\n",
      "        DuplicateKernelError\n",
      "        TimeoutError\n",
      "        COMError\n",
      "        ArgumentError\n",
      "        ReturnValueIgnoredError\n",
      "        KeyReuseError\n",
      "        UnknownKeyError\n",
      "        LeakedCallbackError\n",
      "        BadYieldError\n",
      "        ReturnValueIgnoredError\n",
      "        Return\n",
      "        ErrorDuringImport\n",
      "        NotOneValueFound\n",
      "        CannotEval\n",
      "        OptionError\n",
      "        BdbQuit\n",
      "        Restart\n",
      "        FindCmdError\n",
      "        HomeDirError\n",
      "        ProfileDirError\n",
      "        IPythonCoreError\n",
      "            TryNext\n",
      "            UsageError\n",
      "            StdinNotImplementedError\n",
      "        InputRejected\n",
      "        GetoptError\n",
      "        ErrorToken\n",
      "        PrefilterError\n",
      "        AliasError\n",
      "            InvalidAliasError\n",
      "        Error\n",
      "            InterfaceError\n",
      "            DatabaseError\n",
      "                InternalError\n",
      "                OperationalError\n",
      "                ProgrammingError\n",
      "                IntegrityError\n",
      "                DataError\n",
      "                NotSupportedError\n",
      "        Warning\n",
      "        SpaceInInput\n",
      "        DOMException\n",
      "            IndexSizeErr\n",
      "            DomstringSizeErr\n",
      "            HierarchyRequestErr\n",
      "            WrongDocumentErr\n",
      "            InvalidCharacterErr\n",
      "            NoDataAllowedErr\n",
      "            NoModificationAllowedErr\n",
      "            NotFoundErr\n",
      "            NotSupportedErr\n",
      "            InuseAttributeErr\n",
      "            InvalidStateErr\n",
      "            SyntaxErr\n",
      "            InvalidModificationErr\n",
      "            NamespaceErr\n",
      "            InvalidAccessErr\n",
      "            ValidationErr\n",
      "        ValidationError\n",
      "        EditReadOnlyBuffer\n",
      "        _Retry\n",
      "        InvalidLayoutError\n",
      "        HeightIsUnknownError\n",
      "        ParserSyntaxError\n",
      "        InternalParseError\n",
      "        _PositionUpdatingFinished\n",
      "        SimpleGetItemNotFound\n",
      "        UncaughtAttributeError\n",
      "        HasNoContext\n",
      "        ParamIssue\n",
      "        BadZipFile\n",
      "        LargeZipFile\n",
      "        _JediError\n",
      "            InternalError\n",
      "            WrongVersion\n",
      "            RefactoringError\n",
      "        OnErrorLeaf\n",
      "        InvalidPythonEnvironment\n",
      "        MessageError\n",
      "            MessageParseError\n",
      "                HeaderParseError\n",
      "                    _InvalidEwError\n",
      "                BoundaryError\n",
      "            MultipartConversionError\n",
      "            CharsetError\n",
      "        Error\n",
      "        HTTPException\n",
      "            NotConnected\n",
      "            InvalidURL\n",
      "            UnknownProtocol\n",
      "            UnknownTransferEncoding\n",
      "            UnimplementedFileMode\n",
      "            IncompleteRead\n",
      "            ImproperConnectionState\n",
      "                CannotSendRequest\n",
      "                CannotSendHeader\n",
      "                ResponseNotReady\n",
      "            BadStatusLine\n",
      "                RemoteDisconnected\n",
      "            LineTooLong\n",
      "        InteractivelyDefined\n",
      "        KillEmbedded\n",
      "        Error\n",
      "            NoSuchProcess\n",
      "                ZombieProcess\n",
      "            AccessDenied\n",
      "            TimeoutExpired\n",
      "        Error\n",
      "        TimeoutExpired\n",
      "        TimeoutAbandoned\n",
      "        QueueEmpty\n",
      "        QueueFull\n",
      "        DistutilsError\n",
      "            DistutilsModuleError\n",
      "            DistutilsClassError\n",
      "            DistutilsGetoptError\n",
      "            DistutilsArgError\n",
      "            DistutilsFileError\n",
      "            DistutilsOptionError\n",
      "            DistutilsSetupError\n",
      "            DistutilsPlatformError\n",
      "            DistutilsExecError\n",
      "            DistutilsInternalError\n",
      "            DistutilsTemplateError\n",
      "            DistutilsByteCompileError\n",
      "            RemovedCommandError\n",
      "            PackageDiscoveryError\n",
      "        CCompilerError\n",
      "            PreprocessError\n",
      "            CompileError\n",
      "            LibError\n",
      "            LinkError\n",
      "            UnknownFileError\n",
      "        Error\n",
      "            NoSectionError\n",
      "            DuplicateSectionError\n",
      "            DuplicateOptionError\n",
      "            NoOptionError\n",
      "            InterpolationError\n",
      "                InterpolationMissingOptionError\n",
      "                InterpolationSyntaxError\n",
      "                InterpolationDepthError\n",
      "            ParsingError\n",
      "                MissingSectionHeaderError\n",
      "        ExpatError\n",
      "        error\n",
      "        com_error\n",
      "        internal_error\n",
      "        ParseBaseException\n",
      "            ParseException\n",
      "            ParseFatalException\n",
      "                ParseSyntaxException\n",
      "        RecursiveGrammarException\n",
      "        ResolutionError\n",
      "            VersionConflict\n",
      "                ContextualVersionConflict\n",
      "            DistributionNotFound\n",
      "            UnknownExtra\n",
      "        ParseBaseException\n",
      "            ParseException\n",
      "            ParseFatalException\n",
      "                ParseSyntaxException\n",
      "        RecursiveGrammarException\n",
      "        Error\n",
      "        Unresolved\n",
      "        SkipTest\n",
      "        _ShouldStop\n",
      "        _UnexpectedSuccess\n",
      "        Error\n",
      "            ProtocolError\n",
      "            ResponseError\n",
      "            Fault\n",
      "        UnableToResolveVariableException\n",
      "        InvalidTypeInArgsException\n",
      "    GeneratorExit\n",
      "    SystemExit\n",
      "    KeyboardInterrupt\n",
      "    CancelledError\n",
      "    AbortThread\n",
      "    AbortThread\n",
      "    DebuggerInitializationError\n"
     ]
    }
   ],
   "source": [
    "def print_exception_hierarchy(exception_class, indent=0):\n",
    "    print(\" \" * indent + f\"{exception_class.__name__}\")\n",
    "    for subclass in exception_class.__subclasses__():\n",
    "        print_exception_hierarchy(subclass, indent + 4)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    print(\"Python Exception Hierarchy:\")\n",
    "    print_exception_hierarchy(BaseException)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d7d728f",
   "metadata": {},
   "source": [
    "#### Q3. What errors are defined in the ArithmeticError class? Explain any two with an example."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7192cf4a",
   "metadata": {},
   "source": [
    "\n",
    "The ArithmeticError class is a built-in Python class that serves as a base class for arithmetic-related exceptions. It is used to handle errors that occur during arithmetic operations. Some common errors defined in the ArithmeticError class include:\n",
    "\n",
    "* ZeroDivisionError: This error is raised when trying to divide a number by zero.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a9cd2072",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter number 1 = 10\n",
      "Please enter number 2 = 0\n",
      "division by zero\n",
      "Please enter number 1 = 10\n",
      "Please enter number 2 = 5\n",
      "10/5=2.0\n",
      "_______________________________________________________________________________\n",
      "_______________________________________________________________________________\n"
     ]
    }
   ],
   "source": [
    "# Example 3.1 ZeroDivisionError\n",
    "def divide_number():\n",
    "    num1=int(input(\"Please enter number 1 = \"))\n",
    "    num2=int(input(\"Please enter number 2 = \"))\n",
    "    try:\n",
    "        divided=num1/num2\n",
    "    except Exception as obj:\n",
    "        print(obj)\n",
    "        divide_number()\n",
    "    else:\n",
    "        print(f\"{num1}/{num2}={divided}\")\n",
    "    finally:\n",
    "        print(\"_______________________________________________________________________________\")\n",
    "        \n",
    "divide_number()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ee46dd9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example 3.2 I am unable to find a good example of ArithmeticError (Please suggest one)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bf3c545",
   "metadata": {},
   "source": [
    "#### Q4. Why LookupError class is used? Explain with an example KeyError and IndexError.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dda917da",
   "metadata": {},
   "source": [
    "LookupError is a category of errors in Python that occurs when the code tries to look up or access some value in a data structure (like a dictionary, list, or string) and the value is not found or the index is out of range. The LookupError class is a general base class for these specific types of lookup-related errors.\n",
    "\n",
    "In Python, there are two main types of errors that belong to the LookupError category:\n",
    "* KeyError \n",
    "* IndexError\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "226db349",
   "metadata": {},
   "source": [
    "#### 4.1 KeyError : \n",
    "This error happens when you try to access a key in a dictionary that does not exist. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ecb3283b",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'tomato'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Input \u001b[1;32mIn [18]\u001b[0m, in \u001b[0;36m<cell line: 3>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;66;03m# Example 4.1 :\u001b[39;00m\n\u001b[0;32m      2\u001b[0m fruit_dict \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mapple\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;241m3\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mbanana\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;241m5\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124morange\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;241m2\u001b[39m}\n\u001b[1;32m----> 3\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43mfruit_dict\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtomato\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m)\n",
      "\u001b[1;31mKeyError\u001b[0m: 'tomato'"
     ]
    }
   ],
   "source": [
    "# Example 4.1 :\n",
    "fruit_dict = {'apple': 3, 'banana': 5, 'orange': 2}\n",
    "print(fruit_dict[\"tomato\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4ef14d7",
   "metadata": {},
   "source": [
    "#### 4.2 IndexError: \n",
    "This error happens when you try to access out of indes of the data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "243d5d10",
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Input \u001b[1;32mIn [20]\u001b[0m, in \u001b[0;36m<cell line: 3>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;66;03m# Example 4.2\u001b[39;00m\n\u001b[0;32m      2\u001b[0m lista\u001b[38;5;241m=\u001b[39m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124ma\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mc\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124md\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124me\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m----> 3\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43mlista\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m5\u001b[39;49m\u001b[43m]\u001b[49m)\n",
      "\u001b[1;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "# Example 4.2\n",
    "lista=[\"a\",\"b\",\"c\",\"d\",\"e\"]\n",
    "print(lista[5])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "741dcb76",
   "metadata": {},
   "source": [
    "####  Q5. Explain ImportError. What is ModuleNotFoundError?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14a8c287",
   "metadata": {},
   "source": [
    "ImportError : This error happens when you are trying to access the module that are not found in python"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b12f9585",
   "metadata": {},
   "source": [
    "ImportError is an exception that is raised when there is an error in importing a module or a name in a module. It usually occurs when the Python interpreter cannot find the specified module or when there is a problem with the module's code.\n",
    "\n",
    "ModuleNotFoundError is a specific subclass of ImportError that is raised when the Python interpreter cannot find the specified module. It was introduced in Python 3.6 to provide a more specific error message when a module is not found. Before Python 3.6, an ImportError was raised in such cases.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "34ceea05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Module not found: No module named 'non_existent_module'\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    import non_existent_module\n",
    "except ModuleNotFoundError as e:\n",
    "    print(f\"Module not found: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3fa99a2",
   "metadata": {},
   "source": [
    "#### Q6. List down some best practices for exception handling in python."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65e1636e",
   "metadata": {},
   "source": [
    "Below are all the main points:-\n",
    "\n",
    "1. Use Specific Exceptions: Instead of catching a generic `Exception`, try to catch specific exceptions that you know might be raised. This helps in providing more meaningful error messages and allows you to handle different exceptions differently.\n",
    "\n",
    "2. Avoid Bare Excepts: Avoid using bare `except` statements as they catch all exceptions, including system exit signals. It's better to catch specific exceptions or use `except Exception as e` to capture the exception and handle it appropriately.\n",
    "\n",
    "3. Use `finally`: When necessary, use the `finally` block to ensure that certain code is executed whether an exception is raised or not. For example, closing a file or releasing resources.\n",
    "\n",
    "4. Don't Ignore Exceptions: Never ignore exceptions. At the very least, log the exception so that you can track and diagnose any issues that might occur.\n",
    "\n",
    "5. Handle Exceptions Locally: Only catch exceptions that you can handle locally. If you can't handle an exception properly, it's better to let it propagate up the call stack.\n",
    "\n",
    "6. Use Context Managers: For resources like files, database connections, or network connections, use context managers (`with` statement) to ensure proper cleanup and resource release.\n",
    "\n",
    "7. Avoid Deep Nested `try-except` Blocks: Avoid deeply nesting `try-except` blocks as it can make the code harder to read and maintain. Refactor the code into smaller functions and handle exceptions at appropriate levels.\n",
    "\n",
    "8. Raising Custom Exceptions: Create custom exception classes when you need to raise specific exceptions for your application. This can help in better organization and error reporting.\n",
    "\n",
    "9. Use Exception Chaining: When re-raising an exception, use the `from` keyword to chain exceptions together. This way, you can see the original cause of the exception.\n",
    "\n",
    "10. Keep Error Messages Clear and Informative: When handling exceptions, provide clear and informative error messages. This will help in identifying the issue and debugging the code.\n",
    "\n",
    "Remember, exception handling is an essential part of writing robust and reliable code. Properly handling exceptions can prevent crashes, improve error reporting, and make debugging easier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0259f6a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
